weave/trace/async_job_queue.py

```diff
@@ -0,0 +1,120 @@
+import atexit
+import concurrent.futures
+import threading
+from concurrent.futures import Future
+from typing import Any, Callable, TypeVar
+
+T = TypeVar("T")
+
+MAX_WORKER_DEFAULT = 2**6
+
+
+class AsyncJobQueue:
+    """A queue for managing asynchronous job execution.
+
+    This class provides a thread-safe way to submit and execute jobs asynchronously
+    using a ThreadPoolExecutor. It ensures proper shutdown of the executor when the
+    object is deleted or when the program exits. If jobs are submitted after shutdown,
+    the queue will restart automatically.
+
+    Attributes:
+        executor (concurrent.futures.ThreadPoolExecutor): The executor used to run jobs.
+        _lock (threading.Lock): A lock to ensure thread-safe operations.
+        _is_shutdown (bool): A flag indicating whether the queue has been shut down.
+        _active_jobs (set): A set to keep track of active jobs and callbacks.
+        _max_workers (int): The maximum number of worker threads to use.
+
+    Args:
+        max_workers (int): The maximum number of worker threads to use. Defaults to 5.
+    """
+
+    def __init__(self, max_workers: int = MAX_WORKER_DEFAULT) -> None:
+        self._max_workers = max_workers
+        self._lock = threading.Lock()
+        self._is_shutdown = True
+        self._active_jobs: set[Future] = set()
+        self._start()
+
+    def _start(self) -> None:
+        """Initializes or reinitializes the executor."""
+        with self._lock:
+            if not self._is_shutdown:
+                return
+            self._is_shutdown = False
+            self.executor = concurrent.futures.ThreadPoolExecutor(
+                max_workers=self._max_workers, thread_name_prefix="AsyncJobQueue"
+            )
+            atexit.register(self.shutdown)
+
+    def submit_job(
+        self, func: Callable[..., T], *args: Any, **kwargs: Any
+    ) -> Future[T]:
+        """Submits a job to be executed asynchronously.
+
+        If the queue has been shut down, it will be restarted automatically.
+
+        Args:
+            func: The function to be executed.
+            *args: Positional arguments to pass to the function.
+            **kwargs: Keyword arguments to pass to the function.
+
+        Returns:
+            A Future representing the execution of the job.
+        """
```

Comment:
i feel like an example might be more useful here

Reply:
added